sdk/dynamicconfig/sdk/metric/dynamicextension.go

```diff
@@ -0,0 +1,123 @@
+
+	"go.opentelemetry.io/contrib/sdk/dynamicconfig/internal/metricpattern"
+)
+
+// This file contains extraneous functionality needed to allow per-metric configuration
+// for the Accumulator.
+
+// Extension to accumulator which allows per-metric collection.
+type DynamicExtension struct {
+	lock sync.Mutex
+
+	// List of current schedules.
+	schedules []*pb.ConfigResponse_MetricConfig_Schedule
+
+	// Maps the instrument to most frequent CollectionPeriod of the schedules it matches.
+	// Updated when new config is applied and new instruments are added.
+	instrumentPeriod map[string]pb.ConfigResponse_MetricConfig_Schedule_CollectionPeriod
+}
+
+func NewDynamicExtension() *DynamicExtension {
+	return &DynamicExtension{
+		instrumentPeriod: make(map[string]pb.ConfigResponse_MetricConfig_Schedule_CollectionPeriod),
+	}
+}
+
+// Find period associated with the instrument name. If it is cached in 
+// ext.instrumentPeriod, use that. Otherwise, find the period from the
+// current list of schedules (choosing the most frequent if multiple
+// schedules match. 
+func (ext *DynamicExtension) FindPeriod(name string) pb.ConfigResponse_MetricConfig_Schedule_CollectionPeriod {
+	ext.lock.Lock()
+	defer ext.lock.Unlock()
+
+	// Check if period associated with instrument name is cached. If so return it.
+	if period, ok := ext.instrumentPeriod[name]; ok {
+		return period
+	}
+
+	// Find schedules that matches with instrument name, and return the most
+	// frequent associated CollectionPeriod.
+	var minPeriod pb.ConfigResponse_MetricConfig_Schedule_CollectionPeriod = 0
+	for _, schedule := range ext.schedules {
+		// To match, name must match at least one InclusionPattern and no
+		// ExclusionPatterns.
+		if metricpattern.Matches(name, schedule.InclusionPatterns) &&
+		!metricpattern.Matches(name, schedule.ExclusionPatterns) &&
+		// Check if the CollectionPeriod is the smallest of all those from
+		// matching schedules so far.
+		(minPeriod == 0 || minPeriod > schedule.Period) {
+			minPeriod = schedule.Period
+		}
+	}
+
+	ext.instrumentPeriod[name] = minPeriod
+	return minPeriod
+}
+
+// Clear instrumentPeriod cache. 
+func (ext *DynamicExtension) Clear() {
+	ext.lock.Lock()
+	defer ext.lock.Unlock()
+	ext.instrumentPeriod = make(map[string]pb.ConfigResponse_MetricConfig_Schedule_CollectionPeriod)
+}
+
+func (ext *DynamicExtension) GetSchedules() []*pb.ConfigResponse_MetricConfig_Schedule {
+	ext.lock.Lock()
+	defer ext.lock.Unlock()
+	return ext.schedules
+}
+
+func (ext *DynamicExtension) SetSchedules(schedules []*pb.ConfigResponse_MetricConfig_Schedule) {
+	ext.lock.Lock()
+	defer ext.lock.Unlock()
+	ext.schedules = schedules
+}
+
+// CollectOptions contains optional parameters for Accumulator.Collect().
+type CollectOptions struct {
+	// All instruments with a CollectPeriod included in `periods` should be
+	// exported.
+	Periods []pb.ConfigResponse_MetricConfig_Schedule_CollectionPeriod
```

Comment:
We can add a TODO here to sort Periods, so that we can use binary search in shouldCollect().

Good enough for now because we don't expect to see a lot of periods.

Reply:
Obsolete with int collection period.